datafusion/expr/src/expr_schema.rs

```diff
@@ -37,19 +37,19 @@ use std::sync::Arc;
 /// trait to allow expr to typable with respect to a schema
 pub trait ExprSchemable {
     /// given a schema, return the type of the expr
-    fn get_type<S: ExprSchema>(&self, schema: &S) -> Result<DataType>;
+    fn get_type(&self, schema: &dyn ExprSchema) -> Result<DataType>;
```

Comment:
I had to change the traits to use `dyn` dispatch rather than generics so that the UDF could use the same object (and e.g. not have to create its own copy of these methods for Expr)

I expect this to have 0 performance impact, but I will run the planning benchmarks to be sure if this acceptable

Reply:
I ran 
```
cargo bench --bench sql_planner
```

And the results looked good ( within the noise threshold / reported 1% slower which I don't attribute to this change)